// ngx-tribute/esm5/lib/ngx-tribute.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, EventEmitter, Input, Optional, Output } from '@angular/core';
import { FormControl, FormControlDirective, FormControlName, NgModel } from '@angular/forms';
import Tribute from 'tributejs';
/**
 * @template T
 */
var NgxTributeDirective = /** @class */ (function () {
    function NgxTributeDirective(element, formControlName, formControlDirective, ngModelDirective) {
        this.element = element;
        this.formControlName = formControlName;
        this.formControlDirective = formControlDirective;
        this.ngModelDirective = ngModelDirective;
        this.onMentioned = new EventEmitter();
        this.mentionItemSelected = new EventEmitter();
    }
    Object.defineProperty(NgxTributeDirective.prototype, "control", {
        get: /**
         * @return {?}
         */
        function () {
            return this.implicitFormControl ||
                (this.formControlName && this.formControlName.control) ||
                (this.formControlDirective && this.formControlDirective.control) ||
                (this.ngModelDirective && this.ngModelDirective.control);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgxTributeDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var options = tslib_1.__assign({}, this.options);
        if (this.menuContainer) {
            (/** @type {?} */ (options)).menuContainer = this.menuContainer;
        }
        this.tribute = new Tribute(options);
        this.tribute.attach(this.element.nativeElement);
        this.element.nativeElement.addEventListener('tribute-replaced', function (event) {
            /** @type {?} */
            var value = ['INPUT', 'TEXTAREA'].includes(_this.element.nativeElement.tagName)
                ? _this.element.nativeElement.value
                : _this.element.nativeElement.innerHTML;
            _this.onMentioned.emit(value);
            _this.mentionItemSelected.emit(event.detail.item.original);
            if (_this.control) {
                _this.control.setValue(value);
            }
        });
    };
    /**
     * @return {?}
     */
    NgxTributeDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.tribute) {
            this.tribute.detach(this.element.nativeElement);
        }
    };
    NgxTributeDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxTribute]',
                    exportAs: 'ngx-tribute'
                },] },
    ];
    /** @nocollapse */
    NgxTributeDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FormControlName, decorators: [{ type: Optional }] },
        { type: FormControlDirective, decorators: [{ type: Optional }] },
        { type: NgModel, decorators: [{ type: Optional }] }
    ]; };
    NgxTributeDirective.propDecorators = {
        options: [{ type: Input, args: ['ngxTribute',] }],
        menuContainer: [{ type: Input }],
        implicitFormControl: [{ type: Input }],
        onMentioned: [{ type: Output }],
        mentionItemSelected: [{ type: Output }]
    };
    return NgxTributeDirective;
}());
export { NgxTributeDirective };
if (false) {
    /** @type {?} */
    NgxTributeDirective.prototype.options;
    /** @type {?} */
    NgxTributeDirective.prototype.menuContainer;
    /** @type {?} */
    NgxTributeDirective.prototype.implicitFormControl;
    /** @type {?} */
    NgxTributeDirective.prototype.onMentioned;
    /** @type {?} */
    NgxTributeDirective.prototype.mentionItemSelected;
    /** @type {?} */
    NgxTributeDirective.prototype.tribute;
    /** @type {?} */
    NgxTributeDirective.prototype.element;
    /** @type {?} */
    NgxTributeDirective.prototype.formControlName;
    /** @type {?} */
    NgxTributeDirective.prototype.formControlDirective;
    /** @type {?} */
    NgxTributeDirective.prototype.ngModelDirective;
}
