// ngx-tribute/bundles/ngx-tribute.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('tributejs')) :
    typeof define === 'function' && define.amd ? define('ngx-tribute', ['exports', '@angular/core', '@angular/forms', 'tributejs'], factory) :
    (factory((global['ngx-tribute'] = {}),global.ng.core,global.ng.forms,null));
}(this, (function (exports,core,forms,Tribute) { 'use strict';

    Tribute = Tribute && Tribute.hasOwnProperty('default') ? Tribute['default'] : Tribute;

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /**
     * @template T
     */
    var NgxTributeDirective = (function () {
        function NgxTributeDirective(element, formControlName, formControlDirective, ngModelDirective) {
            this.element = element;
            this.formControlName = formControlName;
            this.formControlDirective = formControlDirective;
            this.ngModelDirective = ngModelDirective;
            this.onMentioned = new core.EventEmitter();
            this.mentionItemSelected = new core.EventEmitter();
        }
        Object.defineProperty(NgxTributeDirective.prototype, "control", {
            get: /**
             * @return {?}
             */ function () {
                return this.implicitFormControl ||
                    (this.formControlName && this.formControlName.control) ||
                    (this.formControlDirective && this.formControlDirective.control) ||
                    (this.ngModelDirective && this.ngModelDirective.control);
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        NgxTributeDirective.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var options = __assign({}, this.options);
                if (this.menuContainer) {
                    ((options)).menuContainer = this.menuContainer;
                }
                this.tribute = new Tribute(options);
                this.tribute.attach(this.element.nativeElement);
                this.element.nativeElement.addEventListener('tribute-replaced', function (event) {
                    /** @type {?} */
                    var value = ['INPUT', 'TEXTAREA'].includes(_this.element.nativeElement.tagName)
                        ? _this.element.nativeElement.value
                        : _this.element.nativeElement.innerHTML;
                    _this.onMentioned.emit(value);
                    _this.mentionItemSelected.emit(event.detail.item.original);
                    if (_this.control) {
                        _this.control.setValue(value);
                    }
                });
            };
        /**
         * @return {?}
         */
        NgxTributeDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                if (this.tribute) {
                    this.tribute.detach(this.element.nativeElement);
                }
            };
        NgxTributeDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[ngxTribute]',
                        exportAs: 'ngx-tribute'
                    },] },
        ];
        /** @nocollapse */
        NgxTributeDirective.ctorParameters = function () {
            return [
                { type: core.ElementRef },
                { type: forms.FormControlName, decorators: [{ type: core.Optional }] },
                { type: forms.FormControlDirective, decorators: [{ type: core.Optional }] },
                { type: forms.NgModel, decorators: [{ type: core.Optional }] }
            ];
        };
        NgxTributeDirective.propDecorators = {
            options: [{ type: core.Input, args: ['ngxTribute',] }],
            menuContainer: [{ type: core.Input }],
            implicitFormControl: [{ type: core.Input }],
            onMentioned: [{ type: core.Output }],
            mentionItemSelected: [{ type: core.Output }]
        };
        return NgxTributeDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxTributeModule = (function () {
        function NgxTributeModule() {
        }
        NgxTributeModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [NgxTributeDirective],
                        exports: [NgxTributeDirective]
                    },] },
        ];
        return NgxTributeModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.NgxTributeModule = NgxTributeModule;
    exports.NgxTributeDirective = NgxTributeDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
